src/parser/core/Module.ts

```diff
@@ -0,0 +1,273 @@
+		if (!cb) { return }
+
+		// QoL filter transforms
+		filter = this.mapFilterEntity(filter, 'to', 'targetID')
+		filter = this.mapFilterEntity(filter, 'by', 'sourceID')
+		if (filter.abilityId) {
+			if (!filter.ability) {
+				filter.ability = {}
+			}
+			filter.ability.guid = filter.abilityId
+			delete filter.abilityId
+		}
+
+		// Make sure events is an array
+		if (!Array.isArray(events)) {
+			events = [events]
+		}
+
+		// Final hook representation
+		const hook = {
+			events,
+			filter,
+			callback: cb.bind(this),
+		}
+
+		// Hook for each of the events
+		events.forEach(event => {
+			let hooks = this._hooks.get(event)
+
+			// Make sure the map has a key for us
+			if (!hooks) {
+				hooks = new Set()
+				this._hooks.set(event, hooks)
+			}
+
+			// Set the hook
+			hooks.add(hook)
+		})
+
+		// Return the hook representation so it can be removed (later)
+		return hook
+	}
+
+	private mapFilterEntity<T extends Event>(
+		filter: Filter<T>,
+		qol: keyof Filter<T>,
+		raw: keyof T,
+	) {
+		if (!filter[qol]) { return filter }
+
+		const _filter = cloneDeep(filter)
+
+		// TODO: Typing on parser req. for some of this stuff
+		switch (_filter[qol]) {
+			case 'player':
+				_filter[raw] = this.parser.player.id
+				break
+			case 'pet':
+				_filter[raw] = this.parser.player.pets.map((pet: any) => pet.id)
+				break
+			default:
+				_filter[raw] = _filter[qol]
+		}
+
+		delete _filter[qol]
+
+		return _filter
+	}
+
+	// TODO: Test
+	protected removeHook(hook: Hook<any>) {
+		hook.events.forEach(event => {
+			const hooks = this._hooks.get(event)
+			if (!hooks) { return }
+			hooks.delete(hook)
+		})
+	}
+
+	triggerEvent(event: Event) {
+		// Run through registered hooks. Avoid calling 'all' on symbols, they're internal stuff.
+		if (typeof event.type !== 'symbol') {
+			this._runHooks(event, this._hooks.get('all'))
+		}
+		this._runHooks(event, this._hooks.get(event.type))
+	}
+
+	private _runHooks(event: Event, hooks?: Set<Hook<Event>>) {
+		if (!hooks) { return }
+		hooks.forEach(hook => {
+			// Check the filter
+			if (!this._filterMatches(event, hook.filter)) {
+				return
+			}
+
+			hook.callback(event)
+		})
+	}
+
+	private _filterMatches(event: Event, filter: Filter<Event>) {
+		const match = Object.keys(filter).every(key => {
+			// If the event doesn't have the key we're looking for, just shortcut out
+			if (!event.hasOwnProperty(key)) {
+				return false
+			}
+
+			const filterVal = filter[key]
+			const eventVal = event[key]
+
+			// FFLogs doesn't use arrays inside events themselves, so I'm using them to handle multiple possible values
+			if (Array.isArray(filterVal)) {
+				return filterVal.includes(eventVal)
+			}
+
+			// If it's an object, I need to dig down. Mostly for the `ability` key
+			if (typeof filterVal === 'object') {
+				return this._filterMatches(eventVal, filterVal)
+			}
+
+			// Basic value check
+			return filterVal === eventVal
+		})
+
+		return match
+	}
+
+	output() {
```

Comment:
`output(): React.ReactChild|false`

otherwise this will require all subclasses to only ever return `boolean`.